src/Solvers/multigrid_solver.jl

```diff
@@ -86,54 +108,144 @@ function MultigridSolver(linear_operation!::Function,
+                                          device_matrix,
+                                          device_x,
+                                          device_b,
+                                          csr_matrix
+                                          )
+
+        return MultigridGPUSolver(arch,
+                                  template_field.grid,
+                                  matrix,
+                                  FT(abstol),
+                                  FT(reltol),
+                                  maxiter,
+                                  x_array,
+                                  b_array,
+                                  amgx_solver
+                                  )
+    end
 end
 
+# @kernel function zero_index!(cuarray :: CuArray)
+#     i = @index(Global, Linear)
+#     cuarray[i] = convert(Int32, cuarray[i] - 1)
+# end
 
-function initialize_matrix(template_field, linear_operator!, args...)
+function initialize_matrix(::CPU, template_field, linear_operator!, args...)
     Nx, Ny, Nz = size(template_field)
     A = spzeros(eltype(template_field.grid), Nx*Ny*Nz, Nx*Ny*Nz)
+    make_column(f) = reshape(interior(f), Nx*Ny*Nz)
 
-    fill_matrix_elements!(A, template_field, linear_operator!, args...)
+    eᵢⱼₖ = similar(template_field)
+    ∇²eᵢⱼₖ = similar(template_field)
+
+    for k = 1:Nz, j in 1:Ny, i in 1:Nx
+        parent(eᵢⱼₖ) .= 0
+        parent(∇²eᵢⱼₖ) .= 0
+        eᵢⱼₖ[i, j, k] = 1
+        fill_halo_regions!(eᵢⱼₖ)
+        linear_operator!(∇²eᵢⱼₖ, eᵢⱼₖ, args...)
+
+        A[:, Ny*Nx*(k-1) + Nx*(j-1) + i] .= make_column(∇²eᵢⱼₖ)
+    end
     
-    return arch_sparse_matrix(architecture(template_field), A)
+    return A
 end
 
-function fill_matrix_elements!(A, template_field, linear_operator!, args...)
+function initialize_matrix(::GPU, template_field, linear_operator!, args...)
     Nx, Ny, Nz = size(template_field)
+    FT = eltype(template_field.grid)
+
     make_column(f) = reshape(interior(f), Nx*Ny*Nz)
 
     eᵢⱼₖ = similar(template_field)
     ∇²eᵢⱼₖ = similar(template_field)
 
+    colptr = CuArray{Int}(undef, Nx*Ny*Nz+1)
+    rowval = CuArray{Int}(undef, 0)
+    nzval  = CuArray{FT}(undef, 0)
+
+    CUDA.@allowscalar colptr[1] = 1
+
     for k = 1:Nz, j in 1:Ny, i in 1:Nx
         parent(eᵢⱼₖ) .= 0
         parent(∇²eᵢⱼₖ) .= 0
-        eᵢⱼₖ[i, j, k] = 1
+        CUDA.@allowscalar eᵢⱼₖ[i, j, k] = 1
         fill_halo_regions!(eᵢⱼₖ)
         linear_operator!(∇²eᵢⱼₖ, eᵢⱼₖ, args...)
-
-        A[:, Ny*Nx*(k-1) + Nx*(j-1) + i] .= make_column(∇²eᵢⱼₖ)
+        count = 0
+        for n = 1:Nz, m in 1:Ny, l in 1:Nx
+            CUDA.@allowscalar if ∇²eᵢⱼₖ[l, m, n] != 0
+                append!(rowval, Ny*Nx*(n-1) + Nx*(m-1) + l)
+                CUDA.@allowscalar append!(nzval, ∇²eᵢⱼₖ[l, m, n])
+                count += 1
+            end
+        end
+        CUDA.@allowscalar colptr[Ny*Nx*(k-1) + Nx*(j-1) + i + 1] = colptr[Ny*Nx*(k-1) + Nx*(j-1) + i] + count
```

Comment:
can you do this on a GPU? If not, I would just do everything on a CPU and in the end copy it in the GPU with `CuArray`, @allowscalar is veeeery slow because you are continuously copying from GPU to CPU

Reply:
Yes very good point. My julia kernel programming isn't good enough to know how to do this because we have a global `count` variable and would need to concatenate the `rowval` and `nzval` arrays together at the end - is there a way to do this? I'm happy to just do so on a CPU

Reply:
Since it's a one-off thing, I think we can do it on the CPU and convert to CuArrays at the end?

Reply:
@simone-silvestri I'm having some trouble with the copying the arrays to CPU implementation due to the call to the `linear_operator!`. For the laplacian we use the [compute_∇²!](https://github.com/CliMA/Oceananigans.jl/blob/13b7f2366726b3e52303dee67bab264941088554/validation/elliptic_solvers/doubly_bounded_poisson.jl#L37-L50) function which takes `arch` as an argument. This means when we convert arrays to the CPU the kernel call still tries to execute on a GPU (as `arch = GPU()`) which rightly throws an error. To keep things generic we parse the arguments to the kernel call with `args...` so are no in general able to edit the `arch` parameter.

For the `HydrostaticFreeSurfaceModel` we have used your code to hardcode the matricies so `@allowscalar` is not required in this case.

Reply:
@simone-silvestri we'll leave the `allowscalar` here... it's only used for the demonstration of the Poisson equation. The implicit free surface solver gives the matrix so it doesn't use the `linear_operation!` to construct the matrix.

@elise-palethorpe let's open an issue after this PR is merged to keep a reminder that this can be further optimized.

Reply:
I would just do something like this
```
function initialize_matrix(::GPU, template_field, linear_operator!, args...)
  cpu_template = #copy field to CPU
  
  A = initialize_matrix(CPU(), cpu_template, linear_operator!, args...) # Here maybe if args... are on the GPU() copy them all to CPU()
  
  return CuSparseMatrixCSC(A)
end
```

Reply:
I don't think that would work because `GPU()` is hiding in `args...`. But I'll copy this in #2737 and we can chat there.